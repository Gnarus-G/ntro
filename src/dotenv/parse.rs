use super::typehint_parser::{ParseTyeHint, TypeHint};

type WithLineNumber<T> = (T, usize);

#[derive(Debug)]
pub struct Variable {
    pub type_hint: Option<WithLineNumber<TypeHint>>,
    pub key: String,
}

impl Variable {
    pub fn is_public(&self) -> bool {
        self.key.starts_with("NEXT_PUBLIC_")
    }
}

pub fn parse_variables(source: &str) -> Vec<&str> {
    return source
        .lines()
        .filter_map(|line| {
            if line.starts_with('#') {
                return None;
            }
            return match line.split('=').collect::<Vec<_>>()[..] {
                [ident, ..] if !ident.is_empty() => Some(ident.trim()),
                _ => None,
            };
        })
        .collect::<Vec<_>>();
}

pub fn parse_variables_with_type_hints(source: &str) -> Vec<Variable> {
    enum Token<'source> {
        LineComment(&'source str, usize),
        Ident(&'source str, usize),
    }

    let mut tokens = source
        .lines()
        .enumerate()
        .filter_map(|(l_num, line)| {
            if line.starts_with('#') {
                return Some(Token::LineComment(line, l_num));
            }
            return match line.split('=').collect::<Vec<_>>()[..] {
                [ident, ..] if !ident.is_empty() => Some(Token::Ident(ident.trim(), l_num)),
                _ => None,
            };
        })
        .peekable();

    let mut vars = Vec::new();

    loop {
        match tokens.next() {
            None => break,
            Some(token) => match (token, tokens.peek()) {
                (Token::LineComment(comment, l_num), Some(Token::Ident(ident, _))) => {
                    let var = Variable {
                        type_hint: comment.into_type_hint().map(|th| (th, l_num)),
                        key: ident.to_string(),
                    };
                    vars.push(var);
                    tokens.next();
                }
                (Token::Ident(ident, _), _) => {
                    vars.push(Variable {
                        type_hint: None,
                        key: ident.to_string(),
                    });
                }

                _ => {}
            },
        };
    }

    vars
}
