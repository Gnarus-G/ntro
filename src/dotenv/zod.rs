use colored::Colorize;
use serde_json::Value;
use std::{
    collections::BTreeMap,
    fmt::Display,
    fs::File,
    io::{BufReader, BufWriter, Write},
    ops::Deref,
    path::{Path, PathBuf},
    sync::Arc,
};

use anyhow::{anyhow, Context, Result};

use crate::command::prettify;

use super::{
    parse::{get_texts, parse_variables_with_type_hints, Variable},
    typehint_parser::TypeHint,
};

use thiserror::Error;

#[derive(Error, Debug)]
pub enum ParseError {
    #[error("\n{a}\n\nconflicts with:\n\n{b}\n")]
    ConflictingTypes { a: TypeHintAt, b: TypeHintAt },
}

#[derive(Debug)]
pub struct TypeHintAt {
    pub th: TypeHint,
    pub line: usize,
    pub meta: Metadata,
}

impl Display for TypeHintAt {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let mut lines = self.meta.source.lines().skip(self.line);

        let curr_line = lines
            .next()
            .expect("assumption that the type hint was parsed along with its line number, failed");
        let next_line = lines
            .next()
            .expect("assumption that the type hint was parsed along with the variable it was decorating, failed");

        writeln!(f, "{}", self.meta.path.to_string_lossy().dimmed())?;
        writeln!(
            f,
            "  {}| {}",
            self.line + 1,
            curr_line
                .replace(
                    self.th.to_string().as_str(),
                    self.th.to_string().green().to_string().as_str()
                )
                .bold()
        )?;
        write!(f, "  {}| {}", self.line + 2, next_line)?;

        Ok(())
    }
}

impl From<(&Metadata, &(TypeHint, usize))> for TypeHintAt {
    fn from(value: (&Metadata, &(TypeHint, usize))) -> Self {
        Self {
            th: value.1 .0.clone(),
            meta: value.0.clone(),
            line: value.1 .1,
        }
    }
}

#[derive(Debug, Clone)]
pub struct Metadata {
    source: Arc<str>,
    path: Arc<Path>,
}

pub fn generate_zod_schema(files: &[PathBuf]) -> Result<String> {
    let text_and_file_names = get_texts(files);

    let sources = text_and_file_names.iter().map(|(source, path)| Metadata {
        source: source.as_str().into(),
        path: path.as_path().into(),
    });

    generate_zod_schema_from_texts(sources)
}

pub fn generate_zod_schema_from_texts(sources: impl Iterator<Item = Metadata>) -> Result<String> {
    let mut map: BTreeMap<String, (Variable, Metadata)> = BTreeMap::new();

    let variables = sources.flat_map(|meta| -> Vec<(Variable, Metadata)> {
        let vars = parse_variables_with_type_hints(meta.source.deref())
            .into_iter()
            .map(|var| (var, meta.clone()))
            .collect();

        return vars;
    });

    for (var, meta) in variables {
        if let Some((v, o_meta)) = map.get(&var.key) {
            if let (Some(lt), Some(rt)) = (&v.type_hint, &var.type_hint) {
                if lt.0 != rt.0 {
                    return Err(ParseError::ConflictingTypes {
                        a: (o_meta, lt).into(),
                        b: (&meta, rt).into(),
                    })
                    .context(
                        "found some conflicting types while parsing variables with type hints",
                    );
                }
            }
        }

        map.insert(var.key.clone(), (var, meta));
    }

    let vars = map.into_values().map(|value| value.0).collect::<Vec<_>>();

    let next_public_vars = vars.iter().filter(|&v| v.is_public()).collect::<Vec<_>>();
    let other_vars = vars.iter().filter(|v| !v.is_public()).collect::<Vec<_>>();

    let to_field_schema = |var: &&Variable| -> String {
        format!(
            r#"    {}: {},"#,
            var.key,
            match &var.type_hint {
                Some(th) => match &th.0 {
                    super::typehint_parser::TypeHint::String => "z.string()".to_string(),
                    super::typehint_parser::TypeHint::Number => "z.coerce.number()".to_string(),
                    super::typehint_parser::TypeHint::Boolean => "z.coerce.boolean()".to_string(),
                    super::typehint_parser::TypeHint::Union(values) =>
                        format!("z.enum([{}])", values.join(",")),
                },
                None => "z.string()".to_string(),
            }
        )
    };

    let js_code = include_str!("module.ts");

    let js_import_line: &str = js_code
        .lines()
        .next()
        .expect("should have an import line at the top of the js implementation");

    let js_impl = js_code
        .lines()
        .skip_while(|line| !line.contains("/* --- MAIN IMPLEMENTATION BELOW --- */"))
        .skip(1)
        .collect::<Vec<_>>()
        .join("\n");

    let output = format!(
        r#"
{js_import_line}

const clientEnvSchemas = {{
{}
}}

const serverEnvSchemas = {{
    ...clientEnvSchemas,
{}
}}

{js_impl}

const processEnv = {{
{}
}}
               "#,
        next_public_vars
            .iter()
            .map(to_field_schema)
            .collect::<Vec<_>>()
            .join("\n"),
        other_vars
            .iter()
            .map(to_field_schema)
            .collect::<Vec<_>>()
            .join("\n"),
        vars.iter()
            .map(|var| format!("   {}: process.env.{},", var.key, var.key))
            .collect::<Vec<_>>()
            .join("\n"),
    );

    Ok(output)
}

pub fn add_tsconfig_path<P: AsRef<Path>>(path: P) -> Result<()> {
    let mut ts_config: Value = File::open("./tsconfig.json")
        .context("couldn't open tsconfig.json")
        .map(BufReader::new)
        .and_then(|reader| serde_json::from_reader(reader).context("failed to parse tsconfig.json"))
        .context("failed to read tsconfig.json")?;

    ts_config
        .get_mut("compilerOptions")
        .context("couldn't find compilerOptions in tsconfig.json")
        .and_then(|paths| {
            paths
                .get_mut("paths")
                .and_then(|node| node.as_object_mut())
                .map(|paths| {
                    paths.insert(
                        "$env".to_string(),
                        Value::Array(vec![Value::String(
                            path.as_ref().to_string_lossy().to_string(),
                        )]),
                    )
                })
                .ok_or(anyhow!("failed to add $env as a path on tsconfig.json"))
        })?;

    File::create("./tsconfig.json")
        .context("failed to open tsconfig.json")
        .map(BufWriter::new)
        .and_then(|mut w| {
            let new_content = serde_json::to_string::<Value>(&ts_config)?;
            let pretty = prettify(new_content.as_bytes(), ".json")?;
            w.write_all(&pretty)
                .context("failed to write to tsconfig.json")?;
            w.flush()
                .context("failed to flush updated tsconfig.json contents")
        })
        .context("failed to update tsconfig.json contents")?;

    Ok(())
}

#[cfg(test)]
mod tests {
    use std::path::{Path, PathBuf};

    use insta::{assert_debug_snapshot, assert_display_snapshot};

    use crate::dotenv::zod::{generate_zod_schema, generate_zod_schema_from_texts};

    #[test]
    fn zod_schema_gen() {
        let output = generate_zod_schema(&[
            PathBuf::from("src/dotenv/.env.test"),
            PathBuf::from("src/dotenv/.env.test2"),
        ])
        .unwrap();
        assert_display_snapshot!(output);
    }

    #[test]
    fn zod_schema_gen_with_type_conflicts() {
        let case = |s: &str| {
            format!(
                r#"
# @type {}
KEY=
            "#,
                s
            )
        };

        fn generate(sources: &[String]) -> Result<String, anyhow::Error> {
            let sources = sources.iter().cloned().enumerate().map(|(i, source)| {
                crate::dotenv::zod::Metadata {
                    source: source.as_str().into(),
                    path: Path::new(&format!("src/dotenv/.env.test.{}", i)).into(),
                }
            });

            generate_zod_schema_from_texts(sources)
        }

        fn gen_err(sources: &[String]) {
            let output = generate(sources).unwrap_err();
            assert_debug_snapshot!(output);
        }

        gen_err(&[case("string"), case("number")]);
        gen_err(&[case("number"), case("boolean")]);
        gen_err(&[case("string"), case("boolean")]);
        gen_err(&[case("'a' | 'b'"), case("number")]);

        gen_err(&[case("string"), case("boolean"), case("number")]);

        // This is not a conflict
        generate(&[case("string"), case("string")]).unwrap();
    }
}
