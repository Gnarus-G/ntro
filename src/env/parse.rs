use chumsky::prelude::*;

#[derive(Debug, Clone, Hash, PartialEq, Eq, PartialOrd, Ord)]
pub struct Variable {
    pub key: String,
}

pub fn parser() -> impl Parser<char, Vec<Variable>, Error = Simple<char>> {
    let comment = just('#')
        .then(take_until(text::newline()))
        .map(|(_, (chars, ..))| String::from_iter(chars));

    let ident = text::ident();

    let value = take_until(text::newline()).map(|(chars, ..)| chars.iter().collect::<String>());

    let line = ident.then(just('=')).then(value).map(|((key, _), _)| {
        return Variable { key };
    });

    line.padded().padded_by(comment.repeated()).repeated()
}
