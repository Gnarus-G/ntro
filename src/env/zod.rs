use chumsky::prelude::*;
use std::{
    fs::File,
    io::{BufReader, Read},
    path::PathBuf,
};

use anyhow::{anyhow, Context, Result};

use super::parse::parser_with_type_hint;

pub fn generate_zod_schema(files: &[PathBuf]) -> Result<String> {
    let parse = |text, file_name| {
        parser_with_type_hint().parse(text).map_err(|err| {
            anyhow!(
                "failed to parse {:?}: {}",
                file_name,
                err.iter()
                    .map(|e| e.to_string())
                    .collect::<Vec<_>>()
                    .join(" ")
            )
        })
    };

    let vars = files
        .iter()
        .map(|file| {
            File::open(file)
                .map(BufReader::new)
                .and_then(|mut rdr| {
                    let mut buf = String::new();
                    rdr.read_to_string(&mut buf).map(|_| buf)
                })
                .context(format!("failed read {file:?}"))
                .and_then(|text| parse(text, file))
        })
        .filter_map(|result| {
            if let Err(e) = &result {
                eprintln!("{e:?}");
            }
            result.ok()
        })
        .flatten()
        .collect::<Vec<_>>();

    let output = format!(
        r#"
import z from "zod";

const envSchema = z.object({{
{}
}})

export const env = envSchema.parse({{
{}
}})
               "#,
        vars.iter()
            .map(|var| format!(
                r#"    {}: {},"#,
                var.key,
                match &var.type_hint {
                    Some(th) => match th {
                        super::typehint_parser::TypeHint::String => "z.string()".to_string(),
                        super::typehint_parser::TypeHint::Number => "z.number()".to_string(),
                        super::typehint_parser::TypeHint::Boolean => "z.boolean()".to_string(),
                        super::typehint_parser::TypeHint::Union(values) =>
                            format!("z.enum([{}])", values.join(",")),
                    },
                    None => "z.string()".to_string(),
                }
            ))
            .collect::<Vec<_>>()
            .join("\n"),
        vars.iter()
            .map(|var| format!("   {}: process.env.{},", var.key, var.key))
            .collect::<Vec<_>>()
            .join("\n")
    );

    Ok(output)
}

#[cfg(test)]
mod tests {
    use std::path::PathBuf;

    use insta::assert_display_snapshot;

    use crate::env::zod::generate_zod_schema;

    #[test]
    fn zod_schema_gen() {
        let output = generate_zod_schema(&[
            PathBuf::from("src/.env.test"),
            PathBuf::from("src/.env.test2"),
        ])
        .unwrap();
        assert_display_snapshot!(output);
    }
}
