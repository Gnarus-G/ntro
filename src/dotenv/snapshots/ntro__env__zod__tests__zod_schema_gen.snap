---
source: src/env/zod.rs
expression: output
---

import z from "zod";

const clientEnvSchemas = {

}

const serverEnvSchemas = {
    NAME: z.string(),
    NAME2: z.enum(['a','b']),
    KEY: z.string(),
    keys: z.string(),
    keys2: z.coerce.string(),
    NAME5: z.string(),
    KEY_Value: z.coerce.number(),
    keys2Da: z.enum(['city','townhall']),
}


export const clientEnv: z.infer<z.ZodObject<typeof clientEnvSchemas>> =
  new Proxy({} as any, {
    get(_, prop: string) {
      try {
        if (prop in clientEnvSchemas) {
          return clientEnvSchemas[prop as keyof typeof clientEnvSchemas].parse(
            processEnv[prop as keyof typeof processEnv],
            { path: [prop] }
          );
        }
        throw new Error(
          `${prop} is not defined for client side environment variables.`
        );
      } catch (e) {
        throw new BadEnvError(`failed to read ${prop} from proccess.env`, e);
      }
    },
  });

export const serverEnv: z.infer<z.ZodObject<typeof serverEnvSchemas>> =
  new Proxy({} as any, {
    get(_, prop: string) {
      try {
        if (prop in serverEnvSchemas) {
          return serverEnvSchemas[prop as keyof typeof serverEnvSchemas].parse(
            processEnv[prop as keyof typeof processEnv],
            { path: [prop] }
          );
        }
        throw new Error(
          `${prop} is not defined for server side environment variables.`
        );
      } catch (e) {
        throw new BadEnvError(`failed to read ${prop} from proccess.env`, e);
      }
    },
  });


const processEnv = {
   NAME: process.env.NAME,
   NAME2: process.env.NAME2,
   KEY: process.env.KEY,
   keys: process.env.keys,
   keys2: process.env.keys2,
   NAME5: process.env.NAME5,
   KEY_Value: process.env.KEY_Value,
   keys2Da: process.env.keys2Da,
}

class BadEnvError extends Error {
    constructor(public message: string, public cause: unknown){
        super(message)
        if (cause instanceof Error) {
          this.message = [message, cause].join("\n â†³ ");
        }
    }
}
               
